<a href="https://colab.research.google.com/github/mayracueva/predicciones_ventas/blob/main/5_Proyecto_1_Parte_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1. Cargar datos

In [ ]:
import pandas as pd
filename = '/content/drive/MyDrive/Data/sales_predictions.csv'
df = pd.read_csv(filename)

#2. Explorar datos

##Ver cabecera del DataFrame

In [ ]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

##Renombrar características

Podemos facilitar el manejo de los datos cambiandole los nombres a las características de  mayúsculas a minúsculas y abreviarlas

In [ ]:
df.rename(columns = {'Item_Identifier' : 'item_id',
                     'Item_Weight' : 'item_w',
                     'Item_Fat_Content' : 'item_fat',
                     'Item_Visibility' : 'item_visib',
                     'Item_Type' : 'item_type',
                     'Item_MRP' : 'item_mrp',
                     'Outlet_Identifier' : 'outlet_id',
                     'Outlet_Establishment_Year' : 'outlet_estab_year',
                     'Outlet_Size' : 'outlet_size',
                     'Outlet_Location_Type' : 'outlet_loc_type',
                     'Outlet_Type' : 'outlet_type',
                     'Item_Outlet_Sales' : 'item_outlet_sales'},
          inplace = True)
df.head()

item_id  item_w item_fat  item_visib              item_type  item_mrp  \
0   FDA15    9.30  Low Fat    0.016047                  Dairy  249.8092   
1   DRC01    5.92  Regular    0.019278            Soft Drinks   48.2692   
2   FDN15   17.50  Low Fat    0.016760                   Meat  141.6180   
3   FDX07   19.20  Regular    0.000000  Fruits and Vegetables  182.0950   
4   NCD19    8.93  Low Fat    0.000000              Household   53.8614   

  outlet_id  outlet_estab_year outlet_size outlet_loc_type        outlet_type  \
0    OUT049               1999      Medium          Tier 1  Supermarket Type1   
1    OUT018               2009      Medium          Tier 3  Supermarket Type2   
2    OUT049               1999      Medium          Tier 1  Supermarket Type1   
3    OUT010               1998         NaN          Tier 3      Grocery Store   
4    OUT013               1987        High          Tier 3  Supermarket Type1   

   item_outlet_sales  
0          3735.1380  
1           443.4228  
2          2097.2700  
3           732.3800  
4           994.7052

##Conocer número de valores y características, tipo de datos y valores no nulos.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   item_id            8523 non-null   object 
 1   item_w             7060 non-null   float64
 2   item_fat           8523 non-null   object 
 3   item_visib         8523 non-null   float64
 4   item_type          8523 non-null   object 
 5   item_mrp           8523 non-null   float64
 6   outlet_id          8523 non-null   object 
 7   outlet_estab_year  8523 non-null   int64  
 8   outlet_size        6113 non-null   object 
 9   outlet_loc_type    8523 non-null   object 
 10  outlet_type        8523 non-null   object 
 11  item_outlet_sales  8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Hay valores faltantes en 'item_w' y 'outlet_size'. A continuación cuantificamos los valores nulos:

##Cuantificar valores nulos

In [ ]:
df.isnull().sum()

item_id                 0
item_w               1463
item_fat                0
item_visib              0
item_type               0
item_mrp                0
outlet_id               0
outlet_estab_year       0
outlet_size          2410
outlet_loc_type         0
outlet_type             0
item_outlet_sales       0
dtype: int64

A 'item_w' le faltan 1463 datos y a 'outlet_size' 2410 datos. Para encontrar el mejor tratamiento para nuestros valores nulos, analizaremos a continuación cada característica:

##Mostrar estadísticas descriptivas

In [ ]:
df.describe().T

count         mean          std       min          25%  \
item_w             7060.0    12.857645     4.643456     4.555     8.773750   
item_visib         8523.0     0.066132     0.051598     0.000     0.026989   
item_mrp           8523.0   140.992782    62.275067    31.290    93.826500   
outlet_estab_year  8523.0  1997.831867     8.371760  1985.000  1987.000000   
item_outlet_sales  8523.0  2181.288914  1706.499616    33.290   834.247400   

                           50%          75%           max  
item_w               12.600000    16.850000     21.350000  
item_visib            0.053931     0.094585      0.328391  
item_mrp            143.012800   185.643700    266.888400  
outlet_estab_year  1999.000000  2004.000000   2009.000000  
item_outlet_sales  1794.331000  3101.296400  13086.964800

El 'item_visib' tiene un valor mínimo de cero. Fuera de eso no se observa nada extraño.

##1. Item_visib

###Ver cabecera con item_visib==0 y valores únicos

In [ ]:
df.loc[df['item_visib'] == 0].head()

item_id  item_w item_fat  item_visib              item_type  item_mrp  \
3    FDX07  19.200  Regular         0.0  Fruits and Vegetables  182.0950   
4    NCD19   8.930  Low Fat         0.0              Household   53.8614   
5    FDP36  10.395  Regular         0.0           Baking Goods   51.4008   
10   FDY07  11.800  Low Fat         0.0  Fruits and Vegetables   45.5402   
32   FDP33  18.700  Low Fat         0.0            Snack Foods  256.6672   

   outlet_id  outlet_estab_year outlet_size outlet_loc_type  \
3     OUT010               1998         NaN          Tier 3   
4     OUT013               1987        High          Tier 3   
5     OUT018               2009      Medium          Tier 3   
10    OUT049               1999      Medium          Tier 1   
32    OUT018               2009      Medium          Tier 3   

          outlet_type  item_outlet_sales  
3       Grocery Store           732.3800  
4   Supermarket Type1           994.7052  
5   Supermarket Type2           556.6088  
10  Supermarket Type1          1516.0266  
32  Supermarket Type2          3068.0064

In [ ]:
df['item_visib'].value_counts()

0.000000    526
0.076975      3
0.162462      2
0.076841      2
0.073562      2
           ... 
0.013957      1
0.110460      1
0.124646      1
0.054142      1
0.044878      1
Name: item_visib, Length: 7880, dtype: int64

Existen 526 valores donde el porcentaje de visualización del producto en la tienda (item_visib) es cero y aún así registra valores de venta (item_outlet_sales). Podría ser que sean datos faltantes y que para no dejar de registrarlo le pusieron cero. No podemos suponer que venden también online ya que de ser el caso, no tendría sentido la variable item_visib y se deberían considerar otras características como tiempo estimado de entrega y precio del costo de envío para la estimación de las ventas.

##2. item_id

###Conocer valores únicos

In [ ]:
df['item_id'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: item_id, Length: 1559, dtype: int64

Existen 1559 valores únicos, es decir, 1559 productos diferentes.

##3. item_w

###Conocer valores únicos

In [ ]:
df['item_w'].value_counts()

12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.275      2
7.685      1
9.420      1
6.520      1
5.400      1
Name: item_w, Length: 415, dtype: int64

###Cuantificar valores nulos

In [ ]:
df['item_w'].isnull().sum()

1463

Como ya se había indicado, 'item_w' tiene 1463 datos perdidos. Como existe la variable 'item_id' que es el código único para cada producto, cada 'item_id' debe tener un 'item_w' único, es decir, cada producto debe tener el mismo peso registrado en 'item_w'. Veamos los datos de estas dos columnas:

###Imputar valores a los nulos

Ver cabecera de item_id y item_w

In [ ]:
df[['item_id', 'item_w']].sort_values(by='item_id').head(20)

item_id  item_w
8043   DRA12   11.60
118    DRA12   11.60
1693   DRA12   11.60
1197   DRA12   11.60
1245   DRA12   11.60
7467   DRA12   11.60
8195   DRA24   19.35
6863   DRA24   19.35
2879   DRA24     NaN
4416   DRA24     NaN
4900   DRA24   19.35
4130   DRA24   19.35
1148   DRA24   19.35
3927   DRA59    8.27
1876   DRA59    8.27
6338   DRA59    8.27
3962   DRA59    8.27
8371   DRA59    8.27
6179   DRA59     NaN
6057   DRA59     NaN

Observamos que en efecto a cada 'item_id' le corresponde un único 'item_w', y en algunos casos hay valores faltantes que podemos recuperar. Para ello vamos a usar la función merge, pero antes debemos generar un nuevo DataFrame con el cual haremos merge:

###Generar DataFrame para merge

In [ ]:
df_w=df.groupby(["item_id"])[["item_w"]].max().reset_index()
df_w.rename(columns = {'item_w' : 'item_w_new'}, inplace = True)
df_w

item_id  item_w_new
0      DRA12      11.600
1      DRA24      19.350
2      DRA59       8.270
3      DRB01       7.390
4      DRB13       6.115
...      ...         ...
1554   NCZ30       6.590
1555   NCZ41      19.850
1556   NCZ42      10.500
1557   NCZ53       9.600
1558   NCZ54      14.650

[1559 rows x 2 columns]

###Hacer merge y ver la cabecera ordenada por 'item_id'

In [ ]:
df_new = pd.merge(left=df, right=df_w, left_on="item_id", right_on="item_id", how="inner")
df_new.sort_values(by='item_id').head(20)

item_id  item_w item_fat  item_visib    item_type  item_mrp outlet_id  \
680    DRA12   11.60  Low Fat    0.041113  Soft Drinks  142.0154    OUT018   
678    DRA12   11.60  Low Fat    0.040912  Soft Drinks  142.3154    OUT013   
677    DRA12   11.60  Low Fat    0.000000  Soft Drinks  141.6154    OUT045   
676    DRA12   11.60  Low Fat    0.041178  Soft Drinks  140.3154    OUT017   
681    DRA12   11.60  Low Fat    0.068535  Soft Drinks  143.0154    OUT010   
679    DRA12   11.60       LF    0.000000  Soft Drinks  141.9154    OUT035   
4801   DRA24   19.35  Regular    0.066832  Soft Drinks  163.8868    OUT010   
4805   DRA24   19.35  Regular    0.039895  Soft Drinks  162.4868    OUT013   
4804   DRA24   19.35  Regular    0.039990  Soft Drinks  165.0868    OUT049   
4800   DRA24     NaN  Regular    0.069909  Soft Drinks  163.2868    OUT019   
4799   DRA24   19.35  Regular    0.040154  Soft Drinks  164.6868    OUT017   
4802   DRA24     NaN  Regular    0.039735  Soft Drinks  165.7868    OUT027   
4803   DRA24   19.35  Regular    0.039921  Soft Drinks  163.3868    OUT035   
6346   DRA59    8.27  Regular    0.214125  Soft Drinks  183.9924    OUT010   
6344   DRA59     NaN  Regular    0.223985  Soft Drinks  186.2924    OUT019   
6343   DRA59     NaN  Regular    0.127308  Soft Drinks  186.6924    OUT027   
6342   DRA59    8.27  Regular    0.000000  Soft Drinks  183.2924    OUT017   
6341   DRA59    8.27  Regular    0.127821  Soft Drinks  185.9924    OUT013   
6340   DRA59    8.27  Regular    0.128127  Soft Drinks  183.6924    OUT049   
6339   DRA59    8.27  Regular    0.127928  Soft Drinks  184.8924    OUT046   

      outlet_estab_year outlet_size outlet_loc_type        outlet_type  \
680                2009      Medium          Tier 3  Supermarket Type2   
678                1987        High          Tier 3  Supermarket Type1   
677                2002         NaN          Tier 2  Supermarket Type1   
676                2007         NaN          Tier 2  Supermarket Type1   
681                1998         NaN          Tier 3      Grocery Store   
679                2004       Small          Tier 2  Supermarket Type1   
4801               1998         NaN          Tier 3      Grocery Store   
4805               1987        High          Tier 3  Supermarket Type1   
4804               1999      Medium          Tier 1  Supermarket Type1   
4800               1985       Small          Tier 1      Grocery Store   
4799               2007         NaN          Tier 2  Supermarket Type1   
4802               1985      Medium          Tier 3  Supermarket Type3   
4803               2004       Small          Tier 2  Supermarket Type1   
6346               1998         NaN          Tier 3      Grocery Store   
6344               1985       Small          Tier 1      Grocery Store   
6343               1985      Medium          Tier 3  Supermarket Type3   
6342               2007         NaN          Tier 2  Supermarket Type1   
6341               1987        High          Tier 3  Supermarket Type1   
6340               1999      Medium          Tier 1  Supermarket Type1   
6339               1997       Small          Tier 1  Supermarket Type1   

      item_outlet_sales  item_w_new  
680            850.8924       11.60  
678           2552.6772       11.60  
677           3829.0158       11.60  
676           2552.6772       11.60  
681            283.6308       11.60  
679            992.7078       11.60  
4801           327.5736       19.35  
4805          4422.2436       19.35  
4804           982.7208       19.35  
4800           491.3604       19.35  
4799          1146.5076       19.35  
4802          4913.6040       19.35  
4803          3439.5228       19.35  
6346           185.0924        8.27  
6344           555.2772        8.27  
6343          7033.5112        8.27  
6342          2406.2012        8.27  
6341           555.2772        8.27  
6340          1295.6468        8.27  
6339          4442.2176        8.27

Al final de la tabla se observa la característica 'item_w_new' la cual contiene los valores del peso de los productos recuperados.

###Conocer número de valores y características, tipo de datos y valores no nulos.

In [ ]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 8522
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   item_id            8523 non-null   object 
 1   item_w             7060 non-null   float64
 2   item_fat           8523 non-null   object 
 3   item_visib         8523 non-null   float64
 4   item_type          8523 non-null   object 
 5   item_mrp           8523 non-null   float64
 6   outlet_id          8523 non-null   object 
 7   outlet_estab_year  8523 non-null   int64  
 8   outlet_size        6113 non-null   object 
 9   outlet_loc_type    8523 non-null   object 
 10  outlet_type        8523 non-null   object 
 11  item_outlet_sales  8523 non-null   float64
 12  item_w_new         8519 non-null   float64
dtypes: float64(5), int64(1), object(7)
memory usage: 932.2+ KB


Ahora tenemos 13 columnas, pero aún con el merge observamos que 'item_w_new' tiene valores perdidos.

###Cuantificar valores nulos

In [ ]:
df_new.isnull().sum()

item_id                 0
item_w               1463
item_fat                0
item_visib              0
item_type               0
item_mrp                0
outlet_id               0
outlet_estab_year       0
outlet_size          2410
outlet_loc_type         0
outlet_type             0
item_outlet_sales       0
item_w_new              4
dtype: int64

Sin embargo, los valores nulos se redujo considerablemente de 1463 a 4. Veamos lo que sucede en estas cuatro filas que contienen los valores perdidos:

###Observar filas con valores perdidos

In [ ]:
item_w_new_missing = df_new['item_w_new'].isna()
df_new.loc[item_w_new_missing, :]

item_id  item_w item_fat  item_visib     item_type  item_mrp outlet_id  \
4125   FDN52     NaN  Regular    0.130933  Frozen Foods   86.9198    OUT027   
6440   FDK57     NaN  Low Fat    0.079904   Snack Foods  120.0440    OUT027   
8278   FDE52     NaN  Regular    0.029742         Dairy   88.9514    OUT027   
8463   FDQ60     NaN  Regular    0.191501  Baking Goods  121.2098    OUT019   

      outlet_estab_year outlet_size outlet_loc_type        outlet_type  \
4125               1985      Medium          Tier 3  Supermarket Type3   
6440               1985      Medium          Tier 3  Supermarket Type3   
8278               1985      Medium          Tier 3  Supermarket Type3   
8463               1985       Small          Tier 1      Grocery Store   

      item_outlet_sales  item_w_new  
4125          1569.9564         NaN  
6440          4434.2280         NaN  
8278          3453.5046         NaN  
8463           120.5098         NaN

Observamos que los valores de 'item_id' correspondientes a los valores faltantes de 'item_w_new' son:
- FDN52
- FDK57
- FDE52
- FDQ60

Para lo cual, vamos a comprobar que del DataFrame original (df) estos valores 'item_id' se repitan una sola vez:

###Contar los valores únicos de 'item_id'

In [ ]:
df_id = df.loc[(df['item_id'] == 'FDN52') | (df['item_id'] == 'FDK57') | (df['item_id'] == 'FDE52') | (df['item_id'] == 'FDQ60'), :]
df_id['item_id'].value_counts()

FDN52    1
FDK57    1
FDE52    1
FDQ60    1
Name: item_id, dtype: int64

En efecto, los valores únicos se repiten una vez, y sus pares correspondientes a la característica 'item_w' es justamente el valor perdido (NaN).

###Calcular porcentaje de valores perdidos

In [ ]:
df_new.isna().sum()/df.shape[0]*100

item_id               0.000000
item_w               17.165317
item_fat              0.000000
item_visib            0.000000
item_type             0.000000
item_mrp              0.000000
outlet_id             0.000000
outlet_estab_year     0.000000
outlet_size          28.276428
outlet_loc_type       0.000000
outlet_type           0.000000
item_outlet_sales     0.000000
item_w_new            0.046932
dtype: float64

###Decidir sobre valores nulos

Dado que no se puede usar otra columna para recuperar estos cuatro datos perdidos ni tampoco estadísticas, y teniendo en cuenta el bajo porcentaje (0.05%< 10%) de nulos que tiene la característica 'item_w_new', podemos tomar la decisión de eliminar las filas que contienen los datos perdidos.

##4. item_fat

###Conocer valores únicos

In [ ]:
df_new['item_fat'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: item_fat, dtype: int64

Al parecer hay un mal registro de valores. Asumimos que 'Low Fat' es lo mismo que 'LF' y que 'low fat'; y 'Regular' es lo mismo que 'reg'

In [ ]:
df_new['item_fat'].replace({'low fat': 'Low Fat', 'LF': 'Low Fat', 'reg': 'Regular'}, inplace = True)
df_new['item_fat'].value_counts()

Low Fat    5517
Regular    3006
Name: item_fat, dtype: int64

##5. item_type

###Conocer valores únicos

In [ ]:
df_new['item_type'].value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: item_type, dtype: int64

Existen 16 valores únicos, es decir, 16 categorías de productos diferentes

##6. item_mrp

###Conocer valores únicos

In [ ]:
df_new['item_mrp'].value_counts()

172.0422    7
142.0154    6
109.5228    6
196.5084    6
188.1872    6
           ..
225.6088    1
223.0088    1
225.5088    1
222.9088    1
108.8938    1
Name: item_mrp, Length: 5938, dtype: int64

##7. outlet_id

###Conocer valores únicos

In [ ]:
df_new['outlet_id'].value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT035    930
OUT046    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: outlet_id, dtype: int64

Existen 10 valores únicos, es decir, 10 tiendas diferentes.

##8. outlet_estab_year

###Conocer valores únicos

In [ ]:
df_new['outlet_estab_year'].value_counts()

1985    1463
1987     932
1999     930
2004     930
1997     930
2002     929
2009     928
2007     926
1998     555
Name: outlet_estab_year, dtype: int64

Existen 9 valores únicos. Dos tiendas se establecieron el mismo año. Lo comprobamos a continuación:

In [ ]:
df_new.groupby(['outlet_id'])[['outlet_estab_year']].max().sort_values(by = 'outlet_estab_year')

outlet_estab_year
outlet_id                   
OUT019                  1985
OUT027                  1985
OUT013                  1987
OUT046                  1997
OUT010                  1998
OUT049                  1999
OUT045                  2002
OUT035                  2004
OUT017                  2007
OUT018                  2009

Las tiendas con el mismo año de establecimiento son:
- OUT019
- OUT027

##9. outlet_size

###Conocer valores únicos

In [ ]:
df_new['outlet_size'].value_counts()

Medium    2793
Small     2388
High       932
Name: outlet_size, dtype: int64

Existen 3 posibles tamaños de tienda.

###Cuantificar valores nulos

In [ ]:
df_new['outlet_size'].isnull().sum()

2410

Como se recuerda, esta característica tiene datos perdidos (2410). Similar a lo que pasaba con el peso de los productos, podemos usar la característica 'outlet_id' para recuperar los valores perdidos. Veamos los datos de estas dos columnas:

###Observar filas con valores perdidos

In [ ]:
outlet_size_missing = df_new['outlet_size'].isna()
df_new.loc[outlet_size_missing, :].head(30)

item_id  item_w item_fat  item_visib              item_type  item_mrp  \
1    FDA15   9.300  Low Fat    0.016055                  Dairy  250.2092   
4    FDA15   9.300  Low Fat    0.026818                  Dairy  248.9092   
7    FDA15   9.300  Low Fat    0.016113                  Dairy  248.8092   
9    DRC01   5.920  Regular    0.019309            Soft Drinks   49.0692   
11   DRC01   5.920  Regular    0.019239            Soft Drinks   49.8692   
15   FDN15  17.500  Low Fat    0.028009                   Meat  141.7180   
18   FDN15  17.500  Low Fat    0.016768                   Meat  141.4180   
21   FDX07  19.200  Regular    0.000000  Fruits and Vegetables  182.0950   
23   FDX07  19.200  Regular    0.022965  Fruits and Vegetables  181.6950   
28   NCD19   8.930  Low Fat    0.022060              Household   56.0614   
29   NCD19   8.930  Low Fat    0.000000              Household   55.4614   
31   NCD19   8.930  Low Fat    0.013254              Household   56.4614   
34   FDP36  10.395  Regular    0.091688           Baking Goods   49.0008   
37   FDO10  13.650  Regular    0.000000            Snack Foods   55.8588   
38   FDO10  13.650  Regular    0.021344            Snack Foods   58.3588   
41   FDO10  13.650  Regular    0.012824            Snack Foods   55.5588   
46   FDP10  19.000  Low Fat    0.128815            Snack Foods  107.6622   
49   FDP10  19.000  Low Fat    0.128350            Snack Foods  106.5622   
51   FDH17  16.200  Regular    0.016687           Frozen Foods   96.9726   
52   FDH17  16.200  Regular    0.016748           Frozen Foods   98.6726   
58   FDU28  19.200  Regular    0.094450           Frozen Foods  187.8214   
64   FDY07  11.800  Low Fat    0.000000  Fruits and Vegetables   46.5402   
65   FDY07  11.800  Low Fat    0.121848  Fruits and Vegetables   46.8402   
66   FDY07  11.800  Low Fat    0.122290  Fruits and Vegetables   46.3402   
73   FDA03  18.500  Regular    0.076097                  Dairy  145.8102   
79   FDX32  15.100  Regular    0.100423  Fruits and Vegetables  144.7786   
80   FDX32  15.100  Regular    0.000000  Fruits and Vegetables  146.2786   
81   FDX32  15.100  Regular    0.100061  Fruits and Vegetables  144.0786   
85   FDS46  17.600  Regular    0.047525            Snack Foods  118.2782   
89   FDF32  16.350  Low Fat    0.068219  Fruits and Vegetables  198.9426   

   outlet_id  outlet_estab_year outlet_size outlet_loc_type  \
1     OUT045               2002         NaN          Tier 2   
4     OUT010               1998         NaN          Tier 3   
7     OUT017               2007         NaN          Tier 2   
9     OUT017               2007         NaN          Tier 2   
11    OUT045               2002         NaN          Tier 2   
15    OUT010               1998         NaN          Tier 3   
18    OUT045               2002         NaN          Tier 2   
21    OUT010               1998         NaN          Tier 3   
23    OUT045               2002         NaN          Tier 2   
28    OUT010               1998         NaN          Tier 3   
29    OUT045               2002         NaN          Tier 2   
31    OUT017               2007         NaN          Tier 2   
34    OUT017               2007         NaN          Tier 2   
37    OUT045               2002         NaN          Tier 2   
38    OUT010               1998         NaN          Tier 3   
41    OUT017               2007         NaN          Tier 2   
46    OUT017               2007         NaN          Tier 2   
49    OUT045               2002         NaN          Tier 2   
51    OUT045               2002         NaN          Tier 2   
52    OUT017               2007         NaN          Tier 2   
58    OUT017               2007         NaN          Tier 2   
64    OUT010               1998         NaN          Tier 3   
65    OUT045               2002         NaN          Tier 2   
66    OUT017               2007         NaN          Tier 2   
73    OUT010               1998         NaN          Tier 3   
79    OU

Los valores que se repiten de 'outlet_id' que corresponden a los valores NaN de 'outlet_size' son:
- OUT045
- OUT010
- OUT017

###Contar los valores únicos de 'outlet_id'

In [ ]:
df_outlet = df_new.loc[(df_new['outlet_id'] == 'OUT045') | (df_new['outlet_id'] == 'OUT010') | (df_new['outlet_id'] == 'OUT017'), :]
print(df_outlet['outlet_id'].value_counts(dropna = False))
df_outlet['outlet_id'].value_counts(dropna = False).sum()

OUT045    929
OUT017    926
OUT010    555
Name: outlet_id, dtype: int64


2410

In [ ]:
print(df_outlet['outlet_id'].value_counts())
df_outlet['outlet_id'].value_counts().sum()

OUT045    929
OUT017    926
OUT010    555
Name: outlet_id, dtype: int64


2410

Se observa que en el conteo de filas considerando valores perdidos como sin considerarlos, la suma es 2410, que es exactamente el número de valores perdidos de nuestro DataFrame para esta característica, lo que nos indica que no podemos recuperar datos perdidos usando 'outlet_id'. También lo podemos ver agrupando las dos características como a continuación:

In [ ]:
df_size1=df_new.groupby(["outlet_id"])[["outlet_size"]].max().reset_index()
df_size1

outlet_id outlet_size
0    OUT010         NaN
1    OUT013        High
2    OUT017         NaN
3    OUT018      Medium
4    OUT019       Small
5    OUT027      Medium
6    OUT035       Small
7    OUT045         NaN
8    OUT046       Small
9    OUT049      Medium

###Usar groupby para ver las combinaciones existentes entre outlet_type, outlet_loc_type y outlet_size

In [ ]:
df_size2=df_new.groupby(["outlet_type", "outlet_loc_type", "outlet_size"])[['item_outlet_sales']].mean()
df_size2

item_outlet_sales
outlet_type       outlet_loc_type outlet_size                   
Grocery Store     Tier 1          Small               340.329723
Supermarket Type1 Tier 1          Medium             2348.354635
                                  Small              2277.844267
                  Tier 2          Small              2438.841866
                  Tier 3          High               2298.995256
Supermarket Type2 Tier 3          Medium             1995.498739
Supermarket Type3 Tier 3          Medium             3694.038558

Observamos que podemos imputar los valores faltantes de outlet_size de la siguiente manera:

1. Las filas que contengan Grocery Store y Tier 1, se le imputará el valor Small.
2. Las filas que contengan Supermarket Type1 y Tier 1, no se le imputará nada porque tiene dos posibilidades (puede ser Medium o Small).
3. Las filas que contengan Supermarket Type1 y Tier 2, se le imputará el valor Small.
4. Las filas que contengan Supermarket Type1 y Tier 3, se le imputará el valor High.
5. Las filas que contengan Supermarket Type2 y Tier 3, se le imputará el valor Medium.
6. Las filas que contengan Supermarket Type3 y Tier 3, se le imputará el valor Medium.

###Imputar valores a los nulos

In [ ]:
for index, row in df_new.iterrows():
  if (row["outlet_type"] == "Grocery Store") & (row["outlet_loc_type"] == "Tier 1"):
    df_new.at[index, "outlet_size_new"] = 'Small'
  elif (row["outlet_type"] == "Supermarket Type1") & (row["outlet_loc_type"] == "Tier 2"):
    df_new.at[index, "outlet_size_new"] = 'Small'
  elif (row["outlet_type"] == "Supermarket Type1") & (row["outlet_loc_type"] == "Tier 3"):
    df_new.at[index, "outlet_size_new"] = 'High'
  elif (row["outlet_type"] == "Supermarket Type2") & (row["outlet_loc_type"] == "Tier 3"):
    df_new.at[index, "outlet_size_new"] = 'Medium'
  elif (row["outlet_type"] == "Supermarket Type3") & (row["outlet_loc_type"] == "Tier 3"):
    df_new.at[index, "outlet_size_new"] = 'Medium'
  else:
    df_new.at[index, "outlet_size_new"] = row["outlet_size"]

###Comparar valores perdidos

In [ ]:
df_new['outlet_size'].isnull().sum()

2410

In [ ]:
df_new['outlet_size_new'].isnull().sum()

555

Se redujo el número de valores perdidos, pasando de 2410 a 555.

In [ ]:
df_size1=df_new.groupby(["outlet_id", "outlet_type", "outlet_loc_type"])[["outlet_size_new"]].count().reset_index()
df_size1

outlet_id        outlet_type outlet_loc_type  outlet_size_new
0    OUT010      Grocery Store          Tier 3                0
1    OUT013  Supermarket Type1          Tier 3              932
2    OUT017  Supermarket Type1          Tier 2              926
3    OUT018  Supermarket Type2          Tier 3              928
4    OUT019      Grocery Store          Tier 1              528
5    OUT027  Supermarket Type3          Tier 3              935
6    OUT035  Supermarket Type1          Tier 2              930
7    OUT045  Supermarket Type1          Tier 2              929
8    OUT046  Supermarket Type1          Tier 1              930
9    OUT049  Supermarket Type1          Tier 1              930

In [ ]:
date_missing = 8523 - df_size1['outlet_size_new'].sum()
date_missing

555

Comprobamos que los 555 datos perdidos corresponden a una sola tienda con número de identificación OUT010, tipo 'Grocery Store' que está ubicada en un área de tipo 'Tier 3'

###Calcular porcentaje de valores perdidos

In [ ]:
df_new.isna().sum()/df_new.shape[0]*100

item_id               0.000000
item_w               17.165317
item_fat              0.000000
item_visib            0.000000
item_type             0.000000
item_mrp              0.000000
outlet_id             0.000000
outlet_estab_year     0.000000
outlet_size          28.276428
outlet_loc_type       0.000000
outlet_type           0.000000
item_outlet_sales     0.000000
item_w_new            0.046932
outlet_size_new       6.511792
dtype: float64

###Decidir sobre valores nulos

Dado que no se puede usar otra columna para recuperar los 555 datos perdidos ni tampoco estadísticas, y aunque el porcentaje de nulos de 'outlet_size_new' es bajo (6.5% < 10%), no podemos tomar la decisión de eliminar las filas que contienen los datos perdidos por dos razones:

1. En este caso en particular, el objetivo del caso es predecir las ventas de los productos por tienda, y si eliminamos las filas que tienen null en 'outlet_size_new' entonces estaríamos perdiendo la oportunidad de predecir las ventas de una tienda.
2. No existen diferencias significativas entre los valores promedio por tamaño de la tienda, pero si podemos observar la variación si se trata por ejemplo de Grocery Store y Supermarket Type3, como se muestra abajo.

In [ ]:
df_size2=df_new.groupby(["outlet_type", "outlet_loc_type", "outlet_size"])[['item_outlet_sales']].mean()
df_size2

item_outlet_sales
outlet_type       outlet_loc_type outlet_size                   
Grocery Store     Tier 1          Small               340.329723
Supermarket Type1 Tier 1          Medium             2348.354635
                                  Small              2277.844267
                  Tier 2          Small              2438.841866
                  Tier 3          High               2298.995256
Supermarket Type2 Tier 3          Medium             1995.498739
Supermarket Type3 Tier 3          Medium             3694.038558

Por tanto, dadas las razones anteriores, se opta por eliminar la característica 'outlet_size_new'.

##10. outlet_loc_type

In [ ]:
df_new['outlet_loc_type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: outlet_loc_type, dtype: int64

Existen 3 posibles tipos de área donde se ubica la tienda.

##11. outlet_type

In [ ]:
df_new['outlet_type'].value_counts()

Supermarket Type1    5577
Grocery Store        1083
Supermarket Type3     935
Supermarket Type2     928
Name: outlet_type, dtype: int64

Existen 4 tipos de tienda.

##12. item_outlet_sales

In [ ]:
df_new['item_outlet_sales'].value_counts()

958.7520     17
1342.2528    16
1845.5976    15
703.0848     15
1230.3984    14
             ..
4102.6596     1
6543.4824     1
3487.4604     1
2066.6432     1
3575.3460     1
Name: item_outlet_sales, Length: 3493, dtype: int64

#3. Eliminar valores nulos

##Eliminar columnas

In [ ]:
df_cln = df_new.drop(columns=['item_w', 'outlet_size', 'outlet_size_new'])
df_cln.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8523 entries, 0 to 8522
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   item_id            8523 non-null   object 
 1   item_fat           8523 non-null   object 
 2   item_visib         8523 non-null   float64
 3   item_type          8523 non-null   object 
 4   item_mrp           8523 non-null   float64
 5   outlet_id          8523 non-null   object 
 6   outlet_estab_year  8523 non-null   int64  
 7   outlet_loc_type    8523 non-null   object 
 8   outlet_type        8523 non-null   object 
 9   item_outlet_sales  8523 non-null   float64
 10  item_w_new         8519 non-null   float64
dtypes: float64(4), int64(1), object(6)
memory usage: 1.0+ MB


##Cambiar nombre a columna

In [ ]:
df_cln = df_cln.rename(columns={'item_w_new' : 'item_w'})
df_cln.head()

item_id item_fat  item_visib item_type  item_mrp outlet_id  \
0   FDA15  Low Fat    0.016047     Dairy  249.8092    OUT049   
1   FDA15  Low Fat    0.016055     Dairy  250.2092    OUT045   
2   FDA15  Low Fat    0.016019     Dairy  248.5092    OUT035   
3   FDA15  Low Fat    0.016088     Dairy  249.6092    OUT018   
4   FDA15  Low Fat    0.026818     Dairy  248.9092    OUT010   

   outlet_estab_year outlet_loc_type        outlet_type  item_outlet_sales  \
0               1999          Tier 1  Supermarket Type1          3735.1380   
1               2002          Tier 2  Supermarket Type1          5976.2208   
2               2004          Tier 2  Supermarket Type1          6474.2392   
3               2009          Tier 3  Supermarket Type2          5976.2208   
4               1998          Tier 3      Grocery Store           498.0184   

   item_w  
0     9.3  
1     9.3  
2     9.3  
3     9.3  
4     9.3

##Eliminar filas con valores nulos

In [ ]:
df_cln.dropna(inplace = True)
df_cln

item_id item_fat  item_visib           item_type  item_mrp outlet_id  \
0      FDA15  Low Fat    0.016047               Dairy  249.8092    OUT049   
1      FDA15  Low Fat    0.016055               Dairy  250.2092    OUT045   
2      FDA15  Low Fat    0.016019               Dairy  248.5092    OUT035   
3      FDA15  Low Fat    0.016088               Dairy  249.6092    OUT018   
4      FDA15  Low Fat    0.026818               Dairy  248.9092    OUT010   
...      ...      ...         ...                 ...       ...       ...   
8518   NCF55  Low Fat    0.021666           Household   33.3874    OUT046   
8519   NCW30  Low Fat    0.011072           Household  257.8962    OUT017   
8520   NCW30  Low Fat    0.011008           Household  259.5962    OUT035   
8521   NCW05  Low Fat    0.148303  Health and Hygiene  108.3938    OUT049   
8522   NCW05  Low Fat    0.148373  Health and Hygiene  108.8938    OUT045   

      outlet_estab_year outlet_loc_type        outlet_type  item_outlet_sales  \
0                  1999          Tier 1  Supermarket Type1          3735.1380   
1                  2002          Tier 2  Supermarket Type1          5976.2208   
2                  2004          Tier 2  Supermarket Type1          6474.2392   
3                  2009          Tier 3  Supermarket Type2          5976.2208   
4                  1998          Tier 3      Grocery Store           498.0184   
...                 ...             ...                ...                ...   
8518               1997          Tier 1  Supermarket Type1          1235.0590   
8519               2007          Tier 2  Supermarket Type1          4402.9354   
8520               2004          Tier 2  Supermarket Type1          2848.9582   
8521               1999          Tier 1  Supermarket Type1          2787.0388   
8522               2002          Tier 2  Supermarket Type1          3644.5892   

      item_w  
0      9.300  
1      9.300  
2      9.300  
3      9.300  
4      9.300  
...      ...  
8518   6.675  
8519   5.210  
8520   5.210  
8521  20.250  
8522  20.250  

[8519 rows x 11 columns]

In [ ]:
df_cln.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8519 entries, 0 to 8522
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   item_id            8519 non-null   object 
 1   item_fat           8519 non-null   object 
 2   item_visib         8519 non-null   float64
 3   item_type          8519 non-null   object 
 4   item_mrp           8519 non-null   float64
 5   outlet_id          8519 non-null   object 
 6   outlet_estab_year  8519 non-null   int64  
 7   outlet_loc_type    8519 non-null   object 
 8   outlet_type        8519 non-null   object 
 9   item_outlet_sales  8519 non-null   float64
 10  item_w             8519 non-null   float64
dtypes: float64(4), int64(1), object(6)
memory usage: 798.7+ KB


Ya tenemos nuestro DataFrame limpio.

#Volver a item_visib

In [ ]:
df_prueba = df_cln.loc[df_cln['item_visib']==0, :]
df_prueba

item_id item_fat  item_visib              item_type  item_mrp outlet_id  \
21     FDX07  Regular         0.0  Fruits and Vegetables  182.0950    OUT010   
24     FDX07  Regular         0.0  Fruits and Vegetables  184.5950    OUT018   
27     NCD19  Low Fat         0.0              Household   53.8614    OUT013   
29     NCD19  Low Fat         0.0              Household   55.4614    OUT045   
32     NCD19  Low Fat         0.0              Household   53.2614    OUT049   
...      ...      ...         ...                    ...       ...       ...   
8473   FDN40  Low Fat         0.0           Frozen Foods  154.2998    OUT046   
8478   FDP15  Low Fat         0.0                   Meat  256.0330    OUT010   
8497   FDH31  Regular         0.0                   Meat   98.2042    OUT019   
8512   FDD28  Low Fat         0.0           Frozen Foods   60.5904    OUT017   
8515   FDU43  Regular         0.0  Fruits and Vegetables  238.6564    OUT013   

      outlet_estab_year outlet_loc_type        outlet_type  item_outlet_sales  \
21                 1998          Tier 3      Grocery Store           732.3800   
24                 2009          Tier 3  Supermarket Type2          3844.9950   
27                 1987          Tier 3  Supermarket Type1           994.7052   
29                 2002          Tier 2  Supermarket Type1           331.5684   
32                 1999          Tier 1  Supermarket Type1           552.6140   
...                 ...             ...                ...                ...   
8473               1997          Tier 1  Supermarket Type1          1999.3974   
8478               1998          Tier 3      Grocery Store          1281.6650   
8497               1985          Tier 1      Grocery Store           198.4084   
8512               2007          Tier 2  Supermarket Type1           410.1328   
8515               1987          Tier 3  Supermarket Type1          3575.3460   

      item_w  
21    19.200  
24    19.200  
27     8.930  
29     8.930  
32     8.930  
...      ...  
8473   5.880  
8478  15.200  
8497  12.000  
8512  10.695  
8515  19.350  

[526 rows x 11 columns]

Pensaría que el item_visib==0 corresponde a un solo tipo de tienda o un solo tipo de producto, pero no es así. Entonces puede que sean valores faltantes que llenaron con cero.

In [ ]:
df_prueba.groupby(["outlet_id", "outlet_type", "outlet_loc_type"])[['item_visib']].max()

item_visib
outlet_id outlet_type       outlet_loc_type            
OUT010    Grocery Store     Tier 3                  0.0
OUT013    Supermarket Type1 Tier 3                  0.0
OUT017    Supermarket Type1 Tier 2                  0.0
OUT018    Supermarket Type2 Tier 3                  0.0
OUT019    Grocery Store     Tier 1                  0.0
OUT027    Supermarket Type3 Tier 3                  0.0
OUT035    Supermarket Type1 Tier 2                  0.0
OUT045    Supermarket Type1 Tier 2                  0.0
OUT046    Supermarket Type1 Tier 1                  0.0
OUT049    Supermarket Type1 Tier 1                  0.0

In [ ]:
df_prueba.groupby(["item_type", "outlet_loc_type"])[['item_visib']].max()

item_visib
item_type             outlet_loc_type            
Baking Goods          Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Breads                Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Breakfast             Tier 1                  0.0
                      Tier 3                  0.0
Canned                Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Dairy                 Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Frozen Foods          Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Fruits and Vegetables Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Hard Drinks           Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Health and Hygiene    Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Household             Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Meat                  Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Others                Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Seafood               Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Snack Foods           Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Soft Drinks           Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0
Starchy Foods         Tier 1                  0.0
                      Tier 2                  0.0
                      Tier 3                  0.0